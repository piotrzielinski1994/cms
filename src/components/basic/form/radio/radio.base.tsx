import { forwardRef, InputHTMLAttributes } from 'react';
import Form from '../root/form';

type InputProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'value' | 'type'> & {
  name: string;
  value?: string;
};

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  return <input ref={ref} type="radio" {...props} />;
});

Input.displayName = 'RadioBase.Input';

const RadioBase = {
  Root: Form.Group,
  Label: Form.Label,
  Input,
  Error: Form.Error,
};

export default RadioBase;
