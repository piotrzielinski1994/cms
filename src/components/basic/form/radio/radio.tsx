import { cn } from '@/utils/tailwind';
import { Circle } from 'lucide-react';
import { ComponentProps, forwardRef } from 'react';
import { checkboxClassNames } from '../checkbox/checkbox';
import RadioBase from './radio.base';

type RadioProps = ComponentProps<typeof RadioBase.Input> & {
  label: string;
  error?: string;
};

const Component = forwardRef<HTMLInputElement, RadioProps>((props, ref) => {
  const { label, error, ...rest } = props;
  return (
    <Root>
      <Label className={checkboxClassNames.wrapper}>
        <Input ref={ref} error={error} {...rest} />
        <span>{label}</span>
      </Label>
      <Error>{error}</Error>
    </Root>
  );
});

const Input = forwardRef<HTMLInputElement, Omit<RadioProps, 'label'>>((props, ref) => {
  const { error, className, ...rest } = props;
  const classNames = {
    wrapper: cn(checkboxClassNames.checkbox({ isValid: !error }), 'rounded-full', className),
    icon: cn(checkboxClassNames.icon, 'w-[0.4lh] h-[0.4lh]', 'fill-current'),
  };

  return (
    <div className={classNames.wrapper}>
      <RadioBase.Input ref={ref} className="sr-only" {...rest} />
      <Circle className={classNames.icon} />
    </div>
  );
});

const Root = RadioBase.Root;
const Label = RadioBase.Label;
const Error = RadioBase.Error;

const Radio = Object.assign(Component, { Root, Label, Input, Error });

export { Radio };
