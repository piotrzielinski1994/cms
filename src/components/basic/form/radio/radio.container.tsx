import { ComponentProps } from 'react';
import { Control, FieldValues, Path, useController } from 'react-hook-form';
import { Radio } from './radio';

type RadioContainerProps<T extends FieldValues> = {
  control: Control<T>;
  name: Path<T>;
} & Omit<ComponentProps<typeof Radio>, 'name'>;

const RadioContainer = <T extends FieldValues>(props: RadioContainerProps<T>) => {
  const { control, name, ...rest } = props;
  const { field, fieldState } = useController({ control, name });
  return <Radio error={fieldState.error?.message} {...rest} {...field} />;
};

export { RadioContainer };
