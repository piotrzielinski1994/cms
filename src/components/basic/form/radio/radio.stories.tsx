import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { useState, type ComponentProps } from 'react';
import { Radio as RadioComponent } from './radio';

type Args = ComponentProps<typeof RadioComponent>;

const meta: Meta<Args> = {
  component: RadioComponent,
  title: 'Components/Basic/Form/Radio',
  argTypes: {
    label: { control: 'text' },
    error: { control: 'text' },
    disabled: { control: 'boolean' },
    name: { table: { disable: true } },
  },
  args: {
    label: DEFAULT_VALUE,
    disabled: false,
  },
};

const Render = ({ label, ...args }: Args) => {
  const t = useTranslations('storybook.basic.form.radio');
  const [selected, setSelected] = useState('');

  const radiosProps = [
    { ...args, label: getFallback(label, t('default.label')) },
    { label: t('disabled.label'), disabled: true },
    { label: t('invalid.label'), error: t('invalid.error') },
  ];

  return (
    <div role="radiogroup" aria-invalid={!!args.error}>
      {radiosProps.map((props, index) => {
        return (
          <RadioComponent
            key={index}
            {...props}
            name="group"
            value={index.toString()}
            checked={selected === index.toString()}
            onChange={(e) => setSelected(e.currentTarget.value)}
          />
        );
      })}
    </div>
  );
};

const Radio: StoryObj<Args> = { render: Render };

export { Radio };
export default meta;
