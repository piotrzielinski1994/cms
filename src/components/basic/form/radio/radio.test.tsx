import { render } from '@testing-library/react';
import { ComponentProps } from 'react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import Form from '../root/form';
import { Radio } from './radio';

describe('Radio', () => {
  const defaultProps = {
    name: 'name',
    id: 'id',
    label: 'Label',
    onChange: vi.fn(),
  } satisfies ComponentProps<typeof Radio>;

  it('should have no accessibility violations', async () => {
    const { container } = render(
      <Form.Group>
        <Form.Label htmlFor={defaultProps.id}>Label</Form.Label>
        <Radio {...defaultProps} />
      </Form.Group>,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const { container } = render(<Radio {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
