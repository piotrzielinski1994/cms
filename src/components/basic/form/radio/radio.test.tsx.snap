// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Radio > should match the snapshot 1`] = `
<div>
  <div
    class="grid gap-1 content-start grid-rows-[auto_1fr]"
  >
    <label
      class="self-end group has-[:disabled]:text-foreground/50 grid grid-cols-[auto_1fr] gap-2 cursor-pointer has-[:disabled]:cursor-not-allowed"
    >
      <div
        class="w-[1lh] h-[1lh] border border-solid border-current bg-input group-hover:group-has-[input:enabled]:border-foreground/90 group-hover:group-has-[input:enabled]:ring-foreground/90 ring-inset tw-has-focus:ring-1 ring-current grid place-items-center rounded-full"
      >
        <input
          class="sr-only"
          id="id"
          name="name"
          type="radio"
        />
        <svg
          aria-hidden="true"
          class="lucide lucide-circle hidden group-has-[:checked]:block w-[0.4lh] h-[0.4lh] fill-current"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
        </svg>
      </div>
      <span>
        Label
      </span>
    </label>
    <span
      class="min-h-[1em] text-sm text-red-500 leading-none"
      role="alert"
    />
  </div>
</div>
`;
